nixos/roles/docker.nix

```diff
@@ -11,40 +12,53 @@ in
-      ip46tables -A fc-docker -m conntrack --ctstate RELATED,ESTABLISHED -j nixos-fw-accept
-      ip46tables -A fc-docker -i ethsrv -j fc-resource-group
-      ip46tables -A fc-docker -i ethsrv -j nixos-fw-refuse
-      ip46tables -A fc-docker -i ethfe -j nixos-fw-refuse
-
-      ip46tables -N DOCKER-USER 2>/dev/null || true
-      ip46tables -I DOCKER-USER 1 ! -i docker+ -o docker+ -j fc-docker
-      # End FC docker rules (1200)
-    '';
-
-    networking.firewall.extraStopCommands = lib.mkOrder 1100 ''
-      # FC docker rules (1100)
-      iptables -w -D nixos-fw -i br-+ -s 172.16.0.0/12 -j nixos-fw-accept 2>/dev/null || true
-      iptables -w -D nixos-fw -i docker+ -s 172.16.0.0/12 -j nixos-fw-accept 2>/dev/null || true
-      ip46tables -D DOCKER-USER ! -i docker+ -o docker+ -j fc-docker 2>/dev/null || true
-      ip46tables -F fc-docker 2>/dev/null || true
-      ip46tables -X fc-docker 2>/dev/null || true
-      # End FC docker rules (1100)
-    '';
+  config = lib.mkMerge [
+    (lib.mkIf (cfg.enable) {
+      environment.systemPackages = [ pkgs.docker-compose ];
+      flyingcircus.users.serviceUsers.extraGroups = [ "docker" ];
+      virtualisation.docker.enable = true;
 
-  };
+      networking.firewall.extraCommands = lib.mkOrder 1200 ''
+        # FC docker rules (1200)
+        # allow access to host from docker networks, we consider this identical
+        # to access from locally running processes.
+        # We grant the full RFC1918 172.16.0.0/12 range.
+        iptables -w -A nixos-fw -i br-+ -s 172.16.0.0/12 -j nixos-fw-accept
+        iptables -w -A nixos-fw -i docker+ -s 172.16.0.0/12 -j nixos-fw-accept
+
+        ip46tables -N fc-docker 2>/dev/null || true
+        ip46tables -A fc-docker -m conntrack --ctstate RELATED,ESTABLISHED -j nixos-fw-accept
+        ip46tables -A fc-docker -i ethsrv -j fc-resource-group
+        ip46tables -A fc-docker -i ethsrv -j nixos-fw-refuse
+        ip46tables -A fc-docker -i ethfe -j nixos-fw-refuse
+
+        ip46tables -N DOCKER-USER 2>/dev/null || true
+        ip46tables -I DOCKER-USER 1 ! -i docker+ -o docker+ -j fc-docker
+        # End FC docker rules (1200)
+      '';
+
+      networking.firewall.extraStopCommands = lib.mkOrder 1100 ''
+        # FC docker rules (1100)
+        iptables -w -D nixos-fw -i br-+ -s 172.16.0.0/12 -j nixos-fw-accept 2>/dev/null || true
+        iptables -w -D nixos-fw -i docker+ -s 172.16.0.0/12 -j nixos-fw-accept 2>/dev/null || true
+        ip46tables -D DOCKER-USER ! -i docker+ -o docker+ -j fc-docker 2>/dev/null || true
+        ip46tables -F fc-docker 2>/dev/null || true
+        ip46tables -X fc-docker 2>/dev/null || true
+        # End FC docker rules (1100)
+      '';
+    })
+
+    (lib.mkIf (!cfg.enable) {
+      networking.firewall.extraStopCommands = lib.mkOrder 1100 ''
+        # FC docker rules (1100) -- ensure absence of docker firewall configuration
+        iptables -D FORWARD -j DOCKER-USER 2>/dev/null || true
```

Comment:
I think we should not remove DOCKER-USER which is created by docker. Users might use docker without our role and add custom config in this chain. Can we just remove the rules we add, as above in `extraStopCommands` for the other case?

Reply:
Hmm? Why are our scripts creating DOCKER-USER in the first place, then?

Reply:
> Can we just remove the rules we add, as above in extraStopCommands for the other case?

Fixed in [04f2249](https://github.com/flyingcircusio/fc-nixos/pull/928/commits/04f224937c854f7baad86a305773c1d37689edaa)?

> Hmm? Why are our scripts creating DOCKER-USER in the first place, then?

Normally the DOCKER-USER chain is created by the docker daemon when it configures its iptables rules, however the docker daemon is only started *after* the NixOS firewall is started. The fc-docker chain which controls access between containers and other hosts in the resource group needs to be called from the DOCKER-USER chain, so both need to be created in the NixOS firewall scripts.